python/ray/experimental/data_loader/shuffle.py

```diff
@@ -0,0 +1,533 @@
+            reducer_part = round_part[reducer_assignment == reducer_idx]
+            reducer_parts.append(reducer_part)
+    if len(reducer_parts) == 1:
+        reducer_parts = reducer_parts[0]
+    duration = timeit.default_timer() - start
+    read_duration = end_read - start
+    if stats_collector is not None:
+        stats_collector.cache_map_done.remote(epoch, duration, read_duration)
+    return reducer_parts
+
+
+def shuffle_from_memory_round(
+        batch_consumer: Callable[[int, Iterable[pd.DataFrame]], None],
+        num_reducers: int,
+        num_trainers: int,
+        epoch: int,
+        round_index: int,
+        trial_start: float,
+        stats_collector: Union[TrialStatsCollectorFromMemory, None],
+        *reducers_partitions: List[List[pd.DataFrame]]) -> None:
+    if stats_collector is not None:
+        stats_collector.round_start.remote(epoch, round_index)
+    shuffled = []
+    # reducers_partitions is a 2D (num_files, max(num_reducers, num_files))
+    # array of reducer partitions; we chunk it into a (num_files, num_reducers)
+    # array of reducer partitions, unzip it into a (num_reducers, num_files)
+    # array of reducer partitions, and feed each set of reducer partitions into
+    # a reducer.
+    for reducer_idx, reducer_partitions in enumerate(zip(
+            *[
+                chunk_it(file_part, num_reducers)
+                for file_part in reducers_partitions])):
+        consumer_batches = shuffle_reduce.options(
+            num_returns=num_trainers).remote(
+                reducer_idx,
+                num_trainers,
+                stats_collector,
+                epoch,
+                round_index,
+                *itertools.chain.from_iterable(reducer_partitions))
+        if not isinstance(consumer_batches, list):
+            consumer_batches = [consumer_batches]
+        shuffled.append(consumer_batches)
+    for trainer_idx, batches in enumerate(
+        zip(
+            *[
+                chunk_it(reducer_part, num_trainers)
+                for reducer_part in shuffled])):
+        consume(
+            trainer_idx, batch_consumer, trial_start, stats_collector, epoch,
+            round_index, list(itertools.chain.from_iterable(batches)))
+    shuffled = list(itertools.chain.from_iterable(shuffled))
+    return shuffled
+
+
+#
+# Shared shuffle stages.
+#
+
+
+@ray.remote
+def shuffle_reduce(
+        reduce_index: int, num_trainers: int,
+        stats_collector: Union[TrialStatsCollector, None], epoch: int,
+        round_index: int, *chunks: pd.DataFrame) -> List[pd.DataFrame]:
+    if stats_collector is not None:
+        stats_collector.reduce_start.remote(epoch, round_index)
+    start = timeit.default_timer()
+    # Concatenate chunks from all mapper partitions.
+    batch = pd.concat(chunks)
+    # Shuffle the batch.
+    batch = batch.sample(frac=1)
+    # Return a list of batches, one for each trainer.
+    batch = np.array_split(batch, num_trainers)
```

Comment:
We should have this output to just one trainer, to avoid having to do another shuffle over the network.